src/drivers/magnetometer/hmc6343/HMC6343.cpp

```diff
@@ -0,0 +1,333 @@
+	ScheduleDelayed(500_ms); // Reset Recovery Time 500 ms
+	
+	uint8_t i2c_address = ReadEEPROM(Register::SLAVE_ADDR);
+
+	if (i2c_address != (I2C_ADDRESS_DEFAULT << 1)) {
+		DEVICE_DEBUG("unexpected I2C slave address 0x%02x", i2c_address);
+		return PX4_ERROR;
+	}
+		
+	// Power-On Start-Up Time 500 ms
+	if (hrt_absolute_time() < 500_ms) {
+		PX4_WARN("Power-On Start-Up Time is 500 ms");
+	}
+
+	uint8_t sn_lsb = ReadEEPROM(Register::SN_LSB);
+	uint8_t sn_msb = ReadEEPROM(Register::SN_MSB);
+	uint16_t serial_num = combine(sn_msb, sn_lsb);
+
+	uint8_t sw_version = ReadEEPROM(Register::SW_VERSION);
+
+	uint8_t date_code_ww = ReadEEPROM(Register::DATE_CODE_WW);
+	uint8_t date_code_yy = ReadEEPROM(Register::DATE_CODE_YY);
+
+	PX4_INFO("Serial Number: 0x%X, Firmware Version: 0x%X, Date (WW): 0x%X, (YY): 0x%X", serial_num, sw_version, date_code_ww, date_code_yy);
+
+	return PX4_OK;
+}
+
+void HMC6343::RunImpl()
+{
+	const hrt_abstime now = hrt_absolute_time();
+
+	switch (_state) {
+	case STATE::RESET:
+
+		// Software Reset
+		SendCommand(Command::RESET);
+		_reset_timestamp = now;
+		_failure_count = 0;
+		_state = STATE::WAIT_FOR_RESET;
+		perf_count(_reset_perf);
+		ScheduleDelayed(500_ms); // Reset Recovery Time 500 ms
+
+		break;
+
+	case STATE::WAIT_FOR_RESET:
+
+		// Check if Run Mode bit is set
+		if ((ReadOPMode(Command::POST_OPMODE1)) & OP_Mode1_BIT::Run) {
+
+			// if reset succeeded then configure
+			_state = STATE::CONFIGURE;
+			ScheduleDelayed(10_ms);
+
+		} else {
+			// RESET not complete
+			if (hrt_elapsed_time(&_reset_timestamp) > 1000_ms) {
+				PX4_DEBUG("Reset failed, retrying");
+				_state = STATE::RESET;
+				ScheduleDelayed(100_ms);
+
+			} else {
+				PX4_DEBUG("Reset not complete, check again in 10 ms");
+				ScheduleDelayed(10_ms);
+			}
+		}
+
+		break;
+
+	case STATE::CONFIGURE:
+
+		if (Configure()) {
+			// if configure succeeded then start read cycle
+			_state = STATE::READ;
+			ScheduleDelayed(200_ms);
+
+		} else {
+			// CONFIGURE not complete
+			if (hrt_elapsed_time(&_reset_timestamp) > 1000_ms) {
+				PX4_DEBUG("Configure failed, resetting");
+				_state = STATE::RESET;
+
+			} else {
+				PX4_DEBUG("Configure failed, retrying");
+			}
+
+			ScheduleDelayed(100_ms);
+		}
+
+		break;
+
+	case STATE::READ: {
+
+			struct TransferBuffer {
+				uint8_t MxMSB;
+				uint8_t MxLSB;
+				uint8_t MyMSB;
+				uint8_t MyLSB;
+				uint8_t MzMSB;
+				uint8_t MzLSB;
+			} buffer{};
+
+			bool success = false;
+			uint8_t cmd = static_cast<uint8_t>(Command::POST_MAG);
+
+			transfer(&cmd, 1, nullptr, 0);
+			ScheduleDelayed(1_ms); // response delay 1 ms
+			if (transfer(nullptr, 0, (uint8_t *)&buffer, sizeof(buffer)) == PX4_OK) {
+
+				int16_t mag_x = combine(buffer.MxMSB, buffer.MxLSB);
+				int16_t mag_y = combine(buffer.MyMSB, buffer.MyLSB);
+				int16_t mag_z = combine(buffer.MzMSB, buffer.MzLSB);
+
+				// sensor's frame is +x forward, +y right, +z up
+				mag_z = (mag_z == INT16_MIN) ? INT16_MAX : -mag_z; // flip z
```

Comment:
Based on the diagram in the datasheet you need to flip y as well.

The data returned to PX4 needs to be +X forward, +Y right, +Z down.

![Screenshot from 2021-05-07 10-08-39](https://user-images.githubusercontent.com/84712/117462042-39499200-af1c-11eb-80ee-a37aa92c1aef.png)